### PR TITLE
WIP: Issue 948 NFS/Avere share mount script

### LINKED ISSUE #948: Write script to mount NFS share from Avere in VM/VMSS
ARM will pass the following to VM's `10.0.1.12-10.0.1.14` 

The script needs to create the following CSV from it `10.0.1.12, 10.0.1.13, 10.0.1.14` and then load balance so that each NFS server receives a similar share of the load. 

This will likely mean adapting [this sample script](https://docs.microsoft.com/en-us/azure/avere-vfxt/avere-vfxt-mount-clients#sample-balanced-client-mounting-script)

### PR DESCRIPTION
**Pull request by lawrencegripper**
Closes #948 

See Issue for notes. 

@martinpeck  as our resident python expert can you take a look?

### CODE REVIEW

File: opencga-app/app/scripts/azure/arm/avere/mount.py [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+def install_apt_package(package):
+    try:
+        print "Attempt to install " + package
+        subprocess.check_call(["apt", "install", package, "-y"])
+        print "Install completed successfully"
+    except subprocess.CalledProcessError as e:
+        print "Failed install " + package + " error:" + e.message
+        exit(4)
+
+def main():
```
Comment (martinpeck) [inline opencga-app/app/scripts/azure/arm/avere/mount.py]: `main` is pretty lengthy

**Comment by lawrencegripper** [comment]
Thanks for the review points I'm going to take a look now. 
Just ran an end2end with the script and it worked as expected on the controller box 

``` text
avere@vm-avere-controller:~$ sudo python test.py avere 10.0.1.12-10.0.1.14
Mounting type:avere
Mounting data:10.0.1.12-10.0.1.14
Attempt to install nfs-common
Reading package lists... Done
Building dependency tree       
Reading state information... Done
nfs-common is already the newest version (1:1.3.4-2.1ubuntu5).
The following packages were automatically installed and are no longer required:
  fakeroot libalgorithm-diff-perl libalgorithm-diff-xs-perl libalgorithm-merge-perl libatomic1 libc-dev-bin libcc1-0 libdpkg-perl libfakeroot
  libfile-fcntllock-perl libgomp1 libisl19 libitm1 liblsan0 libmpc3 libmpx2 libquadmath0 libtsan0 python-all python-asn1crypto python-cffi-backend
  python-cryptography python-dbus python-enum34 python-idna python-ipaddress python-secretstorage
Use 'sudo apt autoremove' to remove them.
0 upgraded, 0 newly installed, 0 to remove and 55 not upgraded.
Install completed successfully
Found ips:10.0.1.12,10.0.1.13,10.0.1.14
Using ip as int: 167772420 for random seed
Primary node selected:10.0.1.14
Secondary nodes selected:10.0.1.12,10.0.1.13
Mounting primary
Mounting secondarys
Done editing fstab ... attempting mount
Attempt #1
avere@vm-avere-controller:~$ cd /media/
primarynfs/    secondarynfs1/ secondarynfs2/ 
avere@vm-avere-controller:~$ cd /media/primarynfs/
avere@vm-avere-controller:/media/primarynfs$ touch 1
avere@vm-avere-controller:/media/primarynfs$ ls /media/secondarynfs1/
1

```

File: opencga-app/app/scripts/azure/arm/avere/mount.py [hunk 2]
```diff
@@ -0,0 +1,201 @@
+import sys
```
[the remark this answers is not in the document]
Comment (lawrencegripper) [reply opencga-app/app/scripts/azure/arm/avere/mount.py]: Good call, done!

File: opencga-app/app/scripts/azure/arm/avere/mount.py [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+# Run `python -m unittest discover` in this dir to execute tests
+
+default_mount_options_nfs = "nfs hard,nointr,proto=tcp,mountproto=tcp,retry=30 0 0"
+default_mount_options_cifs = "dir_mode=0777,file_mode=0777,serverino,nofail,vers=3.0"
+
+
+def get_avere_ips(vserver_string):
+    vserver_range = vserver_string.split("-")
+    if len(vserver_range) < 2:
+        print("Expect vserver ip range to be in the format 'startip-endip' got:" + vserver_string)
```
[the remark this answers is not in the document]
Comment (lawrencegripper) [reply opencga-app/app/scripts/azure/arm/avere/mount.py]: Done

File: opencga-app/app/scripts/azure/arm/avere/mount.py [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+        exit(3)
+
+    all_ips = []
+    start_ip = vserver_range[0]
+    end_ip = vserver_range[1]
+
+    start_ip_parts = start_ip.split(".")
+    start_ip_last_digit = int(start_ip_parts[3])
+    ip_prefix = start_ip_parts[0] + "." + start_ip_parts[1] + "." + start_ip_parts[2] + "."
```
[the remark this answers is not in the document]
Comment (lawrencegripper) [reply opencga-app/app/scripts/azure/arm/avere/mount.py]: Nice, will fix up as like yours more.

File: opencga-app/app/scripts/azure/arm/avere/mount.py [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+    end_ip_last_digit = int(end_ip.split(".")[3])
+
+    for ip in range(start_ip_last_digit, end_ip_last_digit + 1): 
+        all_ips.append(ip_prefix + str(ip))
+    
+    return all_ips
+
+def get_ip_address():
+    s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
```
[the remark this answers is not in the document]
Comment (lawrencegripper) [reply opencga-app/app/scripts/azure/arm/avere/mount.py]: Done as moved to using ipaddress library now

File: opencga-app/app/scripts/azure/arm/avere/mount.py [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+    try:
+        # doesn't even have to be reachable
+        s.connect(('10.255.255.255', 1))
+        IP = s.getsockname()[0]
+    except:
+        IP = '127.0.0.1'
```
[the remark this answers is not in the document]
Comment (lawrencegripper) [reply opencga-app/app/scripts/azure/arm/avere/mount.py]: Agree, will update

File: opencga-app/app/scripts/azure/arm/avere/mount.py [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+    finally:
+        s.close()
+    return IP
+
+def ip_as_int(ip):
+    o = map(int, ip.split('.'))
+    res = (16777216 * o[0]) + (65536 * o[1]) + (256 * o[2]) + o[3]
```
[the remark this answers is not in the document]
Comment (lawrencegripper) [reply opencga-app/app/scripts/azure/arm/avere/mount.py]: This inspired me to try harder and found I could do this `return int(ipaddress.ip_address(ip))`

File: opencga-app/app/scripts/azure/arm/avere/mount.py [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+        print_help()
+        exit(1)
+
+    if mount_data == "":
+        print "Expected second arg to be the mounting data. For avere this is the vserver iprange. Fo azure files this should be the azure files connection details."
+        print_help()
+        exit(2)
+
+
+    print 'Mounting type:' + sys.argv[1]
```
[the remark this answers is not in the document]
Comment (lawrencegripper) [reply opencga-app/app/scripts/azure/arm/avere/mount.py]: Done

File: opencga-app/app/scripts/azure/arm/avere/mount.py (hunk 1 above)
Comment (lawrencegripper) [reply opencga-app/app/scripts/azure/arm/avere/mount.py]: agreed, refactored into azurefiles and avere methods

**Comment by lawrencegripper** [comment]
@martinpeck hopefully all resolved now.

**Comment by martinpeck** [comment]
LGTM!

**Pull request merged**